File: 2022/q06.R
str <- readr::read_file("q06.in")
input <- strsplit(str,"")[[1]]  

#Part 1----
#Add all characters to a vector. If the unique tail (ooooh, I'm a unicorn!) equals 4, jump out of the loop
get_marker <- function(input, no_of_unique_chars){
  characters <- c()
  loop_counter <- 1
  for(char in input){
    characters <- c(characters, char)
    if (length(unique(tail(characters, no_of_unique_chars))) == no_of_unique_chars) return(loop_counter)
    loop_counter <- loop_counter + 1
  }
}
#Answer:
get_marker(input, 4)

#Part 2-----
#Answer:
get_marker(input, 14)

File: 2022/Python/q06.py
with open('q06.in') as file: 
    input = file.read()
# Create a list of the characters in the datastream
datastream = [char for char in input]


# Function to get markers
def get_marker(datastream, no_of_unique_chars):
    for i in (range(0, len(datastream))):
        # Create sets and check if the length == no_of_unique_chars
        if len(set(datastream[i:i+no_of_unique_chars])) == no_of_unique_chars:
            # Answer is i + no_of_unique_chars, as it indicates the step that found the marker
            marker = i + no_of_unique_chars
            break
    return(marker)

answer1 = get_marker(datastream, 4)
answer2 = get_marker(datastream, 14)

print(f"answer1: {answer1}")
print(f"answer1: {answer2}")
